File: src/search_syntax.rs
// 搜索语法解析器 - 支持 Everything 风格的增强语法
use regex::Regex;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone, Default)]
pub struct SearchFilters {
    pub ext: Vec<String>,
    pub size_min: u64,
    pub size_max: u64,
    pub date_after: Option<u64>,
    pub path: String,
    pub name_pattern: String,
}

pub struct SearchSyntaxParser;

impl SearchSyntaxParser {
    pub fn parse(query: &str) -> (String, SearchFilters) {
        let mut filters = SearchFilters::default();
        let mut text = query.to_string();

        // 提取各种语法
        text = Self::extract_ext(&text, &mut filters);
        text = Self::extract_size(&text, &mut filters);
        text = Self::extract_date(&text, &mut filters);
        text = Self::extract_path(&text, &mut filters);
        text = Self::extract_name(&text, &mut filters);

        // 清理多余空格
        let pure_keyword = text.split_whitespace().collect::<Vec<_>>().join(" ");

        (pure_keyword, filters)
    }

    fn extract_ext(text: &str, filters: &mut SearchFilters) -> String {
        let re = Regex::new(r"(?i)ext:([a-zA-Z0-9,]+)").unwrap();
        let result = re.replace_all(text, "");
        
        for cap in re.captures_iter(text) {
            if let Some(m) = cap.get(1) {
                let exts: Vec<String> = m.as_str()
                    .split(',')
                    .filter_map(|e| {
                        let e = e.trim().to_lowercase();
                        if !e.is_empty() { Some(e) } else { None }
                    })
                    .collect();
                filters.ext.extend(exts);
            }
        }
        
        result.to_string()
    }

    fn extract_size(text: &str, filters: &mut SearchFilters) -> String {
        let mut result = text.to_string();

        // size:>100mb
        let re_min = Regex::new(r"(?i)size:>(\d+(?:\.\d+)?)(kb|mb|gb)").unwrap();
        if let Some(cap) = re_min.captures(&result) {
            if let (Some(num), Some(unit)) = (cap.get(1), cap.get(2)) {
                let n: f64 = num.as_str().parse().unwrap_or(0.0);
                filters.size_min = Self::parse_size(n, unit.as_str());
            }
        }
        result = re_min.replace_all(&result, "").to_string();

        // size:<1gb
        let re_max = Regex::new(r"(?i)size:<(\d+(?:\.\d+)?)(kb|mb|gb)").unwrap();
        if let Some(cap) = re_max.captures(&result) {
            if let (Some(num), Some(unit)) = (cap.get(1), cap.get(2)) {
                let n: f64 = num.as_str().parse().unwrap_or(0.0);
                filters.size_max = Self::parse_size(n, unit.as_str());
            }
        }
        result = re_max.replace_all(&result, "").to_string();

        result
    }

    fn parse_size(num: f64, unit: &str) -> u64 {
        let multiplier = match unit.to_lowercase().as_str() {
            "kb" => 1024,
            "mb" => 1024 * 1024,
            "gb" => 1024 * 1024 * 1024,
            _ => 1,
        };
        (num * multiplier as f64) as u64
    }

    fn extract_date(text: &str, filters: &mut SearchFilters) -> String {
        let re = Regex::new(r"(?i)dm:(\S+)").unwrap();
        let result = re.replace_all(text, "");

        if let Some(cap) = re.captures(text) {
            if let Some(date_str) = cap.get(1) {
                let ds = date_str.as_str().to_lowercase();
                let now = SystemTime::now()
                    .duration_since(UNIX_EPOCH)
                    .unwrap()
                    .as_secs();

                let secs_in_day = 86400u64;

                filters.date_after = match ds.as_str() {
                    "today" => Some(now - now % secs_in_day),
                    "yesterday" => Some(now - secs_in_day - now % secs_in_day),
                    "week" => Some(now - 7 * secs_in_day),
                    "month" => Some(now - 30 * secs_in_day),
                    "year" => {
                        // 今年1月1日
                        let year_start = now - (now % (365 * secs_in_day));
                        Some(year_start)
                    }
                    _ => {
                        // 解析相对时间：7d, 12h, 30m
                        let rel_re = Regex::new(r"^(\d+)([dhm])$").unwrap();
                        if let Some(rel_cap) = rel_re.captures(&ds) {
                            if let (Some(num), Some(unit)) = (rel_cap.get(1), rel_cap.get(2)) {
                                let n: u64 = num.as_str().parse().unwrap_or(0);
                                match unit.as_str() {
                                    "d" => Some(now - n * secs_in_day),
                                    "h" => Some(now - n * 3600),
                                    "m" => Some(now - n * 60),
                                    _ => None,
                                }
                            } else {
                                None
                            }
                        } else {
                            None
                        }
                    }
                };
            }
        }

        result.to_string()
    }

    fn extract_path(text: &str, filters: &mut SearchFilters) -> String {
        // path:"C:\Program Files" 或 path:D:\
        let re_quoted = Regex::new(r#"(?i)path:"([^"]+)""#).unwrap();
        if let Some(cap) = re_quoted.captures(text) {
            if let Some(p) = cap.get(1) {
                filters.path = p.as_str().to_string();
            }
            return re_quoted.replace_all(text, "").to_string();
        }

        let re = Regex::new(r"(?i)path:(\S+)").unwrap();
        if let Some(cap) = re.captures(text) {
            if let Some(p) = cap.get(1) {
                filters.path = p.as_str().to_string();
            }
        }
        re.replace_all(text, "").to_string()
    }

    fn extract_name(text: &str, filters: &mut SearchFilters) -> String {
        let re = Regex::new(r"(?i)name:(\S+)").unwrap();
        if let Some(cap) = re.captures(text) {
            if let Some(n) = cap.get(1) {
                filters.name_pattern = n.as_str().to_string();
            }
        }
        re.replace_all(text, "").to_string()
    }

    pub fn apply_filters(results: Vec<crate::commands::SearchResult>, filters: &SearchFilters) -> Vec<crate::commands::SearchResult> {
        results
            .into_iter()
            .filter(|item| Self::match_item(item, filters))
            .collect()
    }

    fn match_item(item: &crate::commands::SearchResult, filters: &SearchFilters) -> bool {
        // 扩展名过滤
        if !filters.ext.is_empty() {
            let ext = std::path::Path::new(&item.filename)
                .extension()
                .and_then(|e| e.to_str())
                .unwrap_or("")
                .to_lowercase();
            if !filters.ext.contains(&ext) {
                return false;
            }
        }

        // 大小过滤
        if filters.size_min > 0 && item.size < filters.size_min {
            return false;
        }
        if filters.size_max > 0 && item.size > filters.size_max {
            return false;
        }

        // 日期过滤
        if let Some(date_after) = filters.date_after {
            if item.mtime < date_after {
                return false;
            }
        }

        // 路径过滤
        if !filters.path.is_empty() {
            let path_lower = item.fullpath.to_lowercase();
            let filter_lower = filters.path.to_lowercase();
            if !path_lower.contains(&filter_lower) {
                return false;
            }
        }

        // 文件名模式过滤
        if !filters.name_pattern.is_empty() {
            let pattern = filters.name_pattern.to_lowercase();
            let filename_lower = item.filename.to_lowercase();
            if !filename_lower.contains(&pattern) {
                return false;
            }
        }

        true
    }
}

File: filesearch/core/search_syntax.py
"""
高级搜索语法解析器
支持 ext:、size:、dm:、path:、name:、dir: 等语法
"""
import re
import datetime
import os


class SearchSyntaxParser:
	"""解析 Everything 风格的搜索语法"""
	
	def __init__(self):
		self.parsed_text = ""
		self.filters = {
			"ext": [],
			"size_min": 0,
			"size_max": 0,
			"date_after": None,
			"path": "",
			"name_pattern": "",
			"dir_pattern": "",
			"content_only": False,
		}
	
	def parse(self, query):
		"""
		解析搜索查询，返回 (纯文本关键词, 过滤条件字典)
		
		示例:
		  "report ext:pdf size:>1mb dm:today path:D:\\Projects"
		  → ("report", {"ext": ["pdf"], "size_min": 1048576, ...})
		"""
		if not query:
			return "", {}
		
		# 重置
		self.parsed_text = query
		self.filters = {
			"ext": [],
			"size_min": 0,
			"size_max": 0,
			"date_after": None,
			"path": "",
			"name_pattern": "",
			"dir_pattern": "",
			"content_only": False,
		}
		
		# 提取各种语法
		self.parsed_text = self._extract_ext(self.parsed_text)
		self.parsed_text = self._extract_size(self.parsed_text)
		self.parsed_text = self._extract_date(self.parsed_text)
		self.parsed_text = self._extract_path(self.parsed_text)
		self.parsed_text = self._extract_name(self.parsed_text)
		self.parsed_text = self._extract_dir(self.parsed_text)
		self.parsed_text = self._extract_content(self.parsed_text)
		
		# 清理多余空格
		self.parsed_text = " ".join(self.parsed_text.split())
		
		return self.parsed_text, self.filters
	
	def _extract_ext(self, text):
		"""提取 ext:pdf 或 ext:jpg,png"""
		pattern = r'ext:([a-zA-Z0-9,]+)'
		matches = re.findall(pattern, text, re.IGNORECASE)
		for match in matches:
			exts = [e.strip().lower() for e in match.split(',') if e.strip()]
			self.filters["ext"].extend(exts)
		return re.sub(pattern, '', text, flags=re.IGNORECASE)
	
	def _extract_size(self, text):
		"""提取 size:>100mb、size:<1kb、size:10mb-50mb"""
		# size:>100mb
		pattern1 = r'size:>(\d+(?:\.\d+)?)(kb|mb|gb)'
		match = re.search(pattern1, text, re.IGNORECASE)
		if match:
			num = float(match.group(1))
			unit = match.group(2).lower()
			self.filters["size_min"] = self._parse_size(num, unit)
			text = re.sub(pattern1, '', text, flags=re.IGNORECASE)
		
		# size:<1kb
		pattern2 = r'size:<(\d+(?:\.\d+)?)(kb|mb|gb)'
		match = re.search(pattern2, text, re.IGNORECASE)
		if match:
			num = float(match.group(1))
			unit = match.group(2).lower()
			self.filters["size_max"] = self._parse_size(num, unit)
			text = re.sub(pattern2, '', text, flags=re.IGNORECASE)
		
		# size:10mb-50mb
		pattern3 = r'size:(\d+(?:\.\d+)?)(kb|mb|gb)-(\d+(?:\.\d+)?)(kb|mb|gb)'
		match = re.search(pattern3, text, re.IGNORECASE)
		if match:
			num1 = float(match.group(1))
			unit1 = match.group(2).lower()
			num2 = float(match.group(3))
			unit2 = match.group(4).lower()
			self.filters["size_min"] = self._parse_size(num1, unit1)
			self.filters["size_max"] = self._parse_size(num2, unit2)
			text = re.sub(pattern3, '', text, flags=re.IGNORECASE)
		
		return text
	
	def _parse_size(self, num, unit):
		"""将大小转换为字节数"""
		multipliers = {'kb': 1024, 'mb': 1024*1024, 'gb': 1024*1024*1024}
		return int(num * multipliers.get(unit, 1))
	
	def _extract_date(self, text):
		"""提取 dm:today、dm:week、dm:7d、dm:2024-12-01"""
		pattern = r'dm:(\S+)'
		match = re.search(pattern, text, re.IGNORECASE)
		if match:
			date_str = match.group(1).lower()
			now = datetime.datetime.now()
			
			if date_str == 'today':
				self.filters["date_after"] = now.replace(hour=0, minute=0, second=0, microsecond=0)
			elif date_str == 'yesterday':
				self.filters["date_after"] = (now - datetime.timedelta(days=1)).replace(hour=0, minute=0, second=0, microsecond=0)
			elif date_str == 'week':
				self.filters["date_after"] = now - datetime.timedelta(days=7)
			elif date_str == 'month':
				self.filters["date_after"] = now - datetime.timedelta(days=30)
			elif date_str == 'year':
				self.filters["date_after"] = now.replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0)
			else:
				# 尝试解析相对时间：7d, 12h, 30m
				rel_match = re.match(r'^(\d+)([dhm])$', date_str)
				if rel_match:
					num = int(rel_match.group(1))
					unit = rel_match.group(2)
					if unit == 'd':
						self.filters["date_after"] = now - datetime.timedelta(days=num)
					elif unit == 'h':
						self.filters["date_after"] = now - datetime.timedelta(hours=num)
					elif unit == 'm':
						self.filters["date_after"] = now - datetime.timedelta(minutes=num)
				else:
					# 尝试解析日期格式 YYYY-MM-DD
					try:
						self.filters["date_after"] = datetime.datetime.strptime(date_str, '%Y-%m-%d')
					except ValueError:
						pass
			
			text = re.sub(pattern, '', text, flags=re.IGNORECASE)
		
		return text
	
	def _extract_path(self, text):
		"""提取 path:D:\\Projects 或 path:"C:\\Program Files" """
		# 带引号的路径
		pattern1 = r'path:"([^"]+)"'
		match = re.search(pattern1, text, re.IGNORECASE)
		if match:
			self.filters["path"] = os.path.normpath(match.group(1))
			return re.sub(pattern1, '', text, flags=re.IGNORECASE)
		
		# 不带引号的路径
		pattern2 = r'path:(\S+)'
		match = re.search(pattern2, text, re.IGNORECASE)
		if match:
			self.filters["path"] = os.path.normpath(match.group(1))
			text = re.sub(pattern2, '', text, flags=re.IGNORECASE)
		
		return text
	
	def _extract_name(self, text):
		"""提取 name:readme 或 name:*.log"""
		pattern = r'name:(\S+)'
		match = re.search(pattern, text, re.IGNORECASE)
		if match:
			self.filters["name_pattern"] = match.group(1)
			text = re.sub(pattern, '', text, flags=re.IGNORECASE)
		return text
	
	def _extract_dir(self, text):
		"""提取 dir:projects"""
		pattern = r'dir:(\S+)'
		match = re.search(pattern, text, re.IGNORECASE)
		if match:
			self.filters["dir_pattern"] = match.group(1)
			text = re.sub(pattern, '', text, flags=re.IGNORECASE)
		return text
	
	def _extract_content(self, text):
		"""提取 content: 前缀"""
		if text.strip().lower().startswith('content:'):
			self.filters["content_only"] = True
			return text[8:].strip()
		return text
	
	def apply_filters(self, results):
		"""
		对搜索结果应用过滤条件
		results: [{"fullpath": "...", "size": ..., "mtime": ..., ...}, ...]
		返回: 过滤后的结果列表
		"""
		if not results:
			return results
		
		filtered = []
		for item in results:
			if not self._match_item(item):
				continue
			filtered.append(item)
		
		return filtered
	
	def _match_item(self, item):
		"""检查单个文件是否匹配所有过滤条件"""
		fullpath = item.get("fullpath", "")
		filename = os.path.basename(fullpath)
		dirname = os.path.dirname(fullpath)
		size = item.get("size", 0)
		mtime = item.get("mtime", 0)
		
		# 扩展名过滤
		if self.filters["ext"]:
			ext = os.path.splitext(filename)[1].lstrip('.').lower()
			if ext not in self.filters["ext"]:
				return False
		
		# 大小过滤
		if self.filters["size_min"] > 0 and size < self.filters["size_min"]:
			return False
		if self.filters["size_max"] > 0 and size > self.filters["size_max"]:
			return False
		
		# 日期过滤
		if self.filters["date_after"]:
			if mtime == 0:
				return False
			file_dt = datetime.datetime.fromtimestamp(mtime)
			if file_dt < self.filters["date_after"]:
				return False
		
		# 路径过滤
		if self.filters["path"]:
			path_filter = self.filters["path"].lower()
			if path_filter not in fullpath.lower():
				return False
		
		# 文件名模式过滤
		if self.filters["name_pattern"]:
			pattern = self.filters["name_pattern"].lower()
			# 支持通配符
			if '*' in pattern or '?' in pattern:
				import fnmatch
				if not fnmatch.fnmatch(filename.lower(), pattern):
					return False
			else:
				if pattern not in filename.lower():
					return False
		
		# 目录名过滤
		if self.filters["dir_pattern"]:
			pattern = self.filters["dir_pattern"].lower()
			if pattern not in dirname.lower():
				return False
		
		return True
